TaskLayer/XLSearchTask/XLSearchTask.cs

```diff
@@ -156,27 +154,80 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
                     {
                         csm.BetaPeptide.ResolveAllAmbiguities();
                     }
+                    csm.ResolveProteinPosAmbiguitiesForXl();
                 }
+
                 var orderedCsmsPerScan = RemoveDuplicateFromCsmsPerScan(csmsPerScan).OrderByDescending(p => p.XLTotalScore).ThenBy(p => p.FullSequence + ((p.BetaPeptide == null) ? "" : p.BetaPeptide.FullSequence)).ToList();
+
                 ListOfCsmsPerMS2ScanParsimony.Add(orderedCsmsPerScan);
             }
-
+            SortCsmsAndCalculateDeltaScores(ListOfCsmsPerMS2ScanParsimony);
+            AssignCrossType(ListOfCsmsPerMS2ScanParsimony);
             var filteredAllPsms = new List<CrosslinkSpectralMatch>();
 
             //For each ms2scan, try to find the best candidate psm from the psms list. Add it into filteredAllPsms
-            //This function is for current usage, this can be replaced with PEP value. 
+            //This function is for current usage, this can be replaced with PEP value.
             foreach (var csmsPerScan in ListOfCsmsPerMS2ScanParsimony)
             {
                 filteredAllPsms.Add(csmsPerScan[0]);
             }
 
             PostXLSearchAnalysisTask postXLSearchAnalysisTask = new PostXLSearchAnalysisTask();
+            postXLSearchAnalysisTask.FileSpecificParameters = this.FileSpecificParameters;
+
+            return postXLSearchAnalysisTask.Run(OutputFolder, dbFilenameList, currentRawFileList, taskId, fileSettingsList, filteredAllPsms.OrderByDescending(p => p.XLTotalScore).ToList(), CommonParameters, XlSearchParameters, proteinList, variableModifications, fixedModifications, localizeableModificationTypes, MyTaskResults);
+        }
+
+        public void SortCsmsAndCalculateDeltaScores(List<List<CrosslinkSpectralMatch>> ListOfCsmsPerMS2Scan)
+        {
+            foreach (var csmsPerScan in ListOfCsmsPerMS2Scan)
+            {
+
+                for (int i = 0; i < csmsPerScan.Count; i++)
+                {
+                    var currentCsm = csmsPerScan[i];
+
+                    if (i == 0 && csmsPerScan.Count == 1)
+                    {
+                        currentCsm.SecondBestXlScore = 0;
+                    }
+                    else if (i == 0)
+                    {
+                        currentCsm.SecondBestXlScore = csmsPerScan[1].XLTotalScore;
+                    }
+                    else
+                    {
+                        currentCsm.SecondBestXlScore = csmsPerScan.First().XLTotalScore;
+                    }
+                }
+            }
+
+            ListOfCsmsPerMS2Scan = ListOfCsmsPerMS2Scan.OrderByDescending(c => c.First().XLTotalScore).ThenByDescending(c => c.First().FullSequence + (c.First().BetaPeptide != null ? c.First().BetaPeptide.FullSequence : "")).ToList();
```

Comment:
This doesn't work; it overwrites the object but doesn't modify the reference. You need to use Sort. This also needs a unit test to make sure the sort worked, because it didn't get caught.

Reply:
changed to return ListList

Reply:
this is repaired and unit tested

Reply:
push!